src/spi/blocking.rs

```diff
@@ -1,108 +1,315 @@
+//! ```
+//! # use embedded_hal::spi::blocking::{SpiBus, SpiBusRead, SpiBusWrite, SpiDevice};
+//! pub struct MyDriver<SPI> {
+//!     spi: SPI,
+//! }
+//!
+//! impl<SPI> MyDriver<SPI>
+//! where
+//!     SPI: SpiDevice,
+//!     SPI::Bus: SpiBus, // or SpiBusRead/SpiBusWrite if you only need to read or only write.
+//! {
+//!     pub fn new(spi: SPI) -> Self {
+//!         Self { spi }
+//!     }
+//!
+//!     pub fn read_foo(&mut self) -> Result<[u8; 2], MyError<SPI::Error>> {
+//!         let mut buf = [0; 2];
+//!
+//!         // `transaction` asserts and deasserts CS for us. No need to do it manually!
+//!         self.spi.transaction(|bus| {
+//!             bus.write(&[0x90])?;
+//!             bus.read(&mut buf)
+//!         }).map_err(MyError::Spi)?;
+//!
+//!         Ok(buf)
+//!     }
+//! }
+//!
+//! #[derive(Copy, Clone, Debug)]
+//! enum MyError<SPI> {
+//!     Spi(SPI),
+//!     // Add other errors for your driver here.
+//! }
+//! ```
+//!
+//! If your device **does not have a CS pin**, use [`SpiBus`] (or [`SpiBusRead`], [`SpiBusWrite`]). This will ensure
+//! your driver has exclusive access to the bus, so no other drivers can interfere. It's not possible to safely share
+//! a bus without CS pins. By requiring [`SpiBus`] you disallow sharing, ensuring correct operation.
+//!
+//! ```
+//! # use embedded_hal::spi::blocking::{SpiBus, SpiBusRead, SpiBusWrite};
+//! pub struct MyDriver<SPI> {
+//!     spi: SPI,
+//! }
+//!
+//! impl<SPI> MyDriver<SPI>
+//! where
+//!     SPI: SpiBus, // or SpiBusRead/SpiBusWrite if you only need to read or only write.
+//! {
+//!     pub fn new(spi: SPI) -> Self {
+//!         Self { spi }
+//!     }
+//!
+//!     pub fn read_foo(&mut self) -> Result<[u8; 2], MyError<SPI::Error>> {
+//!         let mut buf = [0; 2];
+//!         self.spi.write(&[0x90]).map_err(MyError::Spi)?;
+//!         self.spi.read(&mut buf).map_err(MyError::Spi)?;
+//!         Ok(buf)
+//!     }
+//! }
+//!
+//! #[derive(Copy, Clone, Debug)]
+//! enum MyError<SPI> {
+//!     Spi(SPI),
+//!     // Add other errors for your driver here.
+//! }
+//! ```
+//!
+//! If you're (ab)using SPI to **implement other protocols** by bitbanging (WS2812B, onewire, generating arbitrary waveforms...), use [`SpiBus`].
+//! SPI bus sharing doesn't make sense at all in this case. By requiring [`SpiBus`] you disallow sharing, ensuring correct operation.
+//!
+//! # For HAL authors
+//!
+//! HALs **must** implement [`SpiBus`], [`SpiBusRead`] and [`SpiBusWrite`]. Users can combine the bus together with the CS pin (which should
+//! implement [`OutputPin`]) using HAL-independent [`SpiDevice`] implementations such as [`ExclusiveDevice`].
+//!
+//! HALs may additionally implement [`SpiDevice`] to **take advantage of hardware CS management**, which may provide some performance
+//! benefits. (There's no point in a HAL implementing [`SpiDevice`] if the CS management is software-only, this task is better left to
+//! the HAL-independent implementations).
```

Comment:
I do not think HALs should ever implement `SpiDevice`.  The reason is that a HAL has no idea about the execution environment.  But for bus sharing, it needs to make assumptions about some mutex primitive to use.  This will quickly lead to compatibility problems between HALs and "OS" crates.  I'll leave a comment below to detail how I would approach hardware CS support instead (that works with this PR's design as it is!).